File: packages/protocol/src/editor.test.ts
import { describe, expect, test } from 'vitest';

import { initialProductState, project, validateEvent, type RawProductEvent } from './index';

const digest = (character: string) => character.repeat(64);
const rawEditorEvent = (cursor: number, type: string, payload: Record<string, unknown>): RawProductEvent => ({
  schemaVersion: 1,
  eventId: `editor-event-${cursor}`,
  taskId: 'task-1',
  cursor,
  occurredAt: '2026-08-06T12:00:00Z',
  type,
  source: { runtimeId: 'runtime-1' },
  payload,
});

const reference = { findingId: 'finding-1', evidenceId: 'evidence-1', startLine: 4, endLine: 8 };
const openedDraft = {
  id: 'draft-1', path: '/lab/app.go', scopeId: 'scope-1', ownerClientId: 'client-1',
  leaseRevision: 2, baseSha256: digest('a'), baseByteLength: 128, encoding: 'utf-8',
  evidenceReferences: [reference],
};

describe('evidence-aware editor protocol', () => {
  test('projects an audited draft, patch, apply, and verification lifecycle', () => {
    const events = [
      rawEditorEvent(1, 'evidence.committed', { evidence: { id: 'evidence-1', taskId: 'task-1', kind: 'source', summary: 'Affected handler', data: { path: '/lab/app.go' } } }),
      rawEditorEvent(2, 'finding.created', { finding: { id: 'finding-1', title: 'Unsafe handler', severity: 'high', status: 'candidate', confidence: 'high', evidenceIds: ['evidence-1'] } }),
      rawEditorEvent(3, 'editor.draft.opened', { draft: openedDraft }),
      rawEditorEvent(4, 'editor.draft.saved', { draftId: 'draft-1', revision: 1, baseSha256: digest('a'), proposedSha256: digest('b'), proposedByteLength: 144 }),
      rawEditorEvent(5, 'editor.patch.applied', { draftId: 'draft-1', revision: 1, baseSha256: digest('a'), proposedSha256: digest('b'), resultSha256: digest('b'), reviewer: 'client-1' }),
      rawEditorEvent(6, 'editor.patch.verified', { draftId: 'draft-1', revision: 1, verificationId: 'verification-1', success: true, evidenceIds: ['evidence-1'] }),
    ];

    let state = initialProductState();
    for (const raw of events) state = project(state, validateEvent(raw)).state;
    const draft = (state as unknown as { editorDrafts: Record<string, Record<string, unknown>> }).editorDrafts['draft-1'];

    expect(draft).toMatchObject({
      status: 'verified', nextRevision: 2, proposedSha256: digest('b'), resultSha256: digest('b'), reviewer: 'client-1',
      verification: { id: 'verification-1', revision: 1, success: true, evidenceIds: ['evidence-1'] },
      evidenceReferences: [reference],
    });
    expect(state.evidence['evidence-1']).toMatchObject({ summary: 'Affected handler', data: { path: '/lab/app.go' } });
  });

  test('rejects malformed editor audit metadata', () => {
    expect(() => validateEvent(rawEditorEvent(1, 'editor.draft.opened', { draft: { ...openedDraft, baseSha256: 'short' } }))).toThrow('invalid_event');
    expect(() => validateEvent(rawEditorEvent(1, 'editor.draft.opened', { draft: { ...openedDraft, evidenceReferences: [{ ...reference, startLine: 9, endLine: 8 }] } }))).toThrow('invalid_event');
    expect(() => validateEvent(rawEditorEvent(1, 'editor.draft.saved', { draftId: 'draft-1', revision: 0, baseSha256: digest('a'), proposedSha256: digest('b'), proposedByteLength: 144 }))).toThrow('invalid_event');
  });

  test('rejects a draft whose provenance is not committed', () => {
    const opened = validateEvent(rawEditorEvent(1, 'editor.draft.opened', { draft: openedDraft }));
    expect(() => project(initialProductState(), opened)).toThrow('editor_provenance_missing');
  });
});

File: internal/productstate/editor_test.go
package productstate_test

import (
	"encoding/json"
	"strings"
	"testing"

	"cyber-code/internal/productstate"
)

func TestProjectTracksEvidenceAwareEditorLifecycle(t *testing.T) {
	t.Parallel()
	digest := func(character string) string { return strings.Repeat(character, 64) }
	state := productstate.Initial()
	events := []struct {
		eventType string
		payload   any
	}{
		{"evidence.committed", map[string]any{"evidence": map[string]any{"id": "evidence-1", "taskId": "task-1", "kind": "source", "summary": "Affected handler", "data": map[string]any{"path": "/lab/app.go"}}}},
		{"finding.created", map[string]any{"finding": map[string]any{"id": "finding-1", "title": "Unsafe handler", "severity": "high", "status": "candidate", "confidence": "high", "evidenceIds": []string{"evidence-1"}}}},
		{"editor.draft.opened", map[string]any{"draft": map[string]any{
			"id": "draft-1", "path": "/lab/app.go", "scopeId": "scope-1", "ownerClientId": "client-1", "leaseRevision": 2,
			"baseSha256": digest("a"), "baseByteLength": 128, "encoding": "utf-8",
			"evidenceReferences": []any{map[string]any{"findingId": "finding-1", "evidenceId": "evidence-1", "startLine": 4, "endLine": 8}},
		}}},
		{"editor.draft.saved", map[string]any{"draftId": "draft-1", "revision": 1, "baseSha256": digest("a"), "proposedSha256": digest("b"), "proposedByteLength": 144}},
		{"editor.patch.applied", map[string]any{"draftId": "draft-1", "revision": 1, "baseSha256": digest("a"), "proposedSha256": digest("b"), "resultSha256": digest("b"), "reviewer": "client-1"}},
		{"editor.patch.verified", map[string]any{"draftId": "draft-1", "revision": 1, "verificationId": "verification-1", "success": true, "evidenceIds": []string{"evidence-1"}}},
	}
	for index, item := range events {
		result, err := productstate.Project(state, mustEvent(t, index+1, item.eventType, item.payload, nil))
		if err != nil {
			t.Fatalf("project %s: %v", item.eventType, err)
		}
		state = result.State
	}
	data, err := json.Marshal(state)
	if err != nil {
		t.Fatal(err)
	}
	var snapshot struct {
		EditorDrafts map[string]struct {
			Status       string `json:"status"`
			NextRevision int    `json:"nextRevision"`
			ResultSHA256 string `json:"resultSha256"`
			Reviewer     string `json:"reviewer"`
			Verification struct {
				ID      string `json:"id"`
				Success bool   `json:"success"`
			} `json:"verification"`
		} `json:"editorDrafts"`
	}
	if err := json.Unmarshal(data, &snapshot); err != nil {
		t.Fatal(err)
	}
	draft := snapshot.EditorDrafts["draft-1"]
	if draft.Status != "verified" || draft.NextRevision != 2 || draft.ResultSHA256 != digest("b") || draft.Reviewer != "client-1" || draft.Verification.ID != "verification-1" || !draft.Verification.Success {
		t.Fatalf("editor draft state = %+v", draft)
	}
}

func TestProjectRejectsMissingEditorProvenance(t *testing.T) {
	t.Parallel()
	_, err := productstate.Project(productstate.Initial(), mustEvent(t, 1, "editor.draft.opened", map[string]any{"draft": map[string]any{
		"id": "draft-1", "path": "/lab/app.go", "scopeId": "scope-1", "ownerClientId": "client-1", "leaseRevision": 1,
		"baseSha256": strings.Repeat("a", 64), "baseByteLength": 128, "encoding": "utf-8",
		"evidenceReferences": []any{map[string]any{"findingId": "finding-1", "evidenceId": "evidence-1", "startLine": 4, "endLine": 8}},
	}}, nil))
	if err == nil || err.Error() != "editor_provenance_missing" {
		t.Fatalf("missing provenance error = %v", err)
	}
}
